TOMLUnmarshaler without the actual TOML object in file causes a nil ptr dereference

**Describe the bug**
Unmarshaling into a struct containing a field that implements `UnmarshalTOML` will cause a nil pointer dereference panic if the TOML object is not there in the tree.

**To Reproduce**

<details>
<summary>Go code to reproduce (or <a href="https://play.golang.org/p/pFhlhqksYgE">playground</a>)</summary>

```go
type Config struct {
	Key string `toml:"key"`
	Obj Custom `toml:"obj"`
}

type Custom struct {
	v string
}

func (c *Custom) UnmarshalTOML(v interface{}) error {
	return nil
}

const t = `key = "value"`

func main() {
	t, err := toml.LoadBytes([]byte(t))
	if err != nil {
		panic(err)
	}

	var c Config

	if err := t.Unmarshal(&c); err != nil {
		panic(err)
	}

	fmt.Printf("%#v\n", c)
}
```

</details>

<details>
<summary>Panic stack trace</summary>

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x503368]

goroutine 1 [running]:
github.com/pelletier/go-toml.(*Tree).ToMap(0x0, 0x56f600)
	/home/diamond/.go/pkg/mod/github.com/pelletier/go-toml@v1.8.0/tomltree_write.go:502 +0x48
github.com/pelletier/go-toml.(*Decoder).valueFromTree(0xc000163f10, 0x56f600, 0x523720, 0x0, 0xc00012c2c0, 0xc00012c210, 0x199, 0xa, 0x10, 0xc0001360c8)
	/home/diamond/.go/pkg/mod/github.com/pelletier/go-toml@v1.8.0/marshal.go:720 +0x112
github.com/pelletier/go-toml.(*Decoder).valueFromTree(0xc000163f10, 0x56f600, 0x526660, 0xc0001183c0, 0xc000163e18, 0x199, 0xc00013c240, 0x0, 0xc000163e30, 0x40bae6)
	/home/diamond/.go/pkg/mod/github.com/pelletier/go-toml@v1.8.0/marshal.go:826 +0x887
github.com/pelletier/go-toml.(*Decoder).unmarshal(0xc000163f10, 0x511b60, 0xc00012c200, 0xc0001183c0, 0x0)
	/home/diamond/.go/pkg/mod/github.com/pelletier/go-toml@v1.8.0/marshal.go:698 +0x275
github.com/pelletier/go-toml.(*Tree).Unmarshal(...)
	/home/diamond/.go/pkg/mod/github.com/pelletier/go-toml@v1.8.0/marshal.go:585
main.main()
	/tmp/tomltest/main.go:39 +0x143
exit status 2
```

</details>

**Expected behavior**
Unmarshaler should not crash and instead should just not call `UnmarshalTOML` and leave the value as it is.

**Versions**
 - go-toml: version 1.8.0
 - go: version 1.14.6
 - operating system: linux/amd64 on NixOS

**Additional context**
A workaround to this bug would be to check if the TOML object exists, and insert an empty one if it does not. The code below fixes it when placed right before `Unmarshal`:

```go
if !t.Has("obj") {
	t.Set("obj", Custom{})
}
```

--- a/marshal.go
+++ b/marshal.go
@@ -742,6 +742,10 @@ func (d *Decoder) valueFromTree(mtype reflect.Type, tval *Tree, mval1 *reflect.V
 	if mvalPtr := reflect.New(mtype); isCustomUnmarshaler(mvalPtr.Type()) {
 		d.visitor.visitAll()
 
+		if tval == nil {
+			return mvalPtr.Elem(), nil
+		}
+
 		if err := callCustomUnmarshaler(mvalPtr, tval.ToMap()); err != nil {
 			return reflect.ValueOf(nil), fmt.Errorf("unmarshal toml: %v", err)
 		}

--- a/marshal_test.go
+++ b/marshal_test.go
@@ -3941,3 +3941,38 @@ bar = 42
 
 	reflect.DeepEqual(x, expected)
 }
+
+type Config struct {
+	Key string `toml:"key"`
+	Obj Custom `toml:"obj"`
+}
+
+type Custom struct {
+	v string
+}
+
+func (c *Custom) UnmarshalTOML(v interface{}) error {
+	c.v = "called"
+	return nil
+}
+
+func TestGithubIssue431(t *testing.T) {
+	doc := `key = "value"`
+	tree, err := LoadBytes([]byte(doc))
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+
+	var c Config
+	if err := tree.Unmarshal(&c); err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+
+	if c.Key != "value" {
+		t.Errorf("expected c.Key='value', not '%s'", c.Key)
+	}
+
+	if c.Obj.v == "called" {
+		t.Errorf("UnmarshalTOML should not have been called")
+	}
+}

